src/FractionalWrapper.sol

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.13;
+
+import "lib/yield-utils-v2/contracts/mocks/ERC20Mock.sol";
+import "lib/yield-utils-v2/contracts/token/IERC20.sol";
+import "lib/openzeppelin-contracts/contracts/access/Ownable.sol";
+
+/**
+@title Fractional Wrapper
+@author Calnix
+@dev Contract allows users to exchange a pre-specified ERC20 token for some other wrapped ERC20 tokens.
+@notice Wrapped tokens will be burned, when user withdraws their deposited tokens.
+*/
+
+contract FractionalWrapper is ERC20Mock, Ownable {
+
+    ///@dev Exchange rate at inception: 1 underlying (DAI) == 1 share (yvDAI) | Ex-rate: 1 DAI/yvDAI = 0.5 -> 1 DAI gets you 1/2 yvDAI
+    uint exRate = 1e27;
+
+    ///@notice ERC20 interface specifying token contract functions
+    ///@dev For constant variables, the value has to be fixed at compile-time, while for immutable, it can still be assigned at construction time.
+    IERC20 public immutable asset;    
+    
+    /// @notice Emit event when ERC20 tokens are deposited into Fractionalized Wrapper
+    event Deposit(address indexed caller, address indexed owner, uint256 assets, uint256 shares);
+    
+    /// @notice Emit event when ERC20 tokens are withdrawn from Fractionalized Wrapper
+    event Withdraw(address indexed caller, address indexed receiver, address indexed owner, uint256 assets, uint256 shares);
+
+    ///@notice Creates a new wrapper token for a specified token 
+    ///@dev Token will have 18 decimal places as ERC20Mock inherits from ERC20Permit
+    ///@param asset_ Address of underlying ERC20 token (e.g. DAI)
+    ///@param tokenName Name of FractionalWrapper tokens 
+    ///@param tokenSymbol Symbol of FractionalWrapper tokens (e.g. yvDAI)
+    constructor(IERC20 asset_, string memory tokenName, string memory tokenSymbol) ERC20Mock(tokenName, tokenSymbol) {
+        asset = asset_;
+    }
+
+
+    /// @notice User to deposit underlying tokens for FractionalWrapper tokens
+    /// @dev Mints shares Vault shares to receiver by depositing exactly amount of underlying tokens
+    /// @param assets The amount of underlying tokens to deposit
+    /// @param receiver Address of receiver of Fractional Wrapper shares
+    function deposit(uint assets, address receiver) external returns(uint shares) {       
+        receiver = msg.sender;
+        shares = convertToShares(assets);
```

Comment:
Find out how to enable gas tracing in foundry and check the result of this call.

We generally avoid calling public functions from the same contract for cost reasons. Instead, we have internal functions with the logic, which are called by all public functions. [Example](https://github.com/yieldprotocol/vault-v2/blob/master/contracts/oracles/chainlink/ChainlinkMultiOracle.sol).

Reply:
Yup, changed both converts to internal functions, that does break the last leg of my testing though. Reached out to Richie for guidance.